### DISCUSSION TITLE
How can one use avsc with uWebSockets.js

### DISCUSSION THREAD

**Discussion by Wyzix33** [uNetworking/uWebSockets.js#715] [Q&A] [answered]
Does anybody have any basic example on how to use uWS with avsc (https://github.com/mtth/avsc) ? 
I'm trying to start a new project for a game and would like to experiment a little.
As I understand i need to create a Schema on the nodejs server to encode my JSON before sending and use that same Schema on browser to decode it back to a JSON...
I was thinking on using uWS to send the schema but it seems they have a service that does that...
I just need a simple example on how to start uWS with avsc and send a JSON from server to client and reverse encoding and decoding it and also dealing with that Schema thing...
I'm using webpack
Thanks

**Comment by e3dio** [comment, marked as answer]
Packing binary data was discussed here https://github.com/uNetworking/uWebSockets.js/discussions/501#discussioncomment-622168 with example using avsc as part of benchmark. Ideally do custom binary encode but if you need something to help you can do `avscSchema=avsc.Type.forValue(data)` and `buffer=avscSchema.toBuffer(data)` `ws.send(buffer,true)`

```javascript
// encode data benchmark
const data={field1:true,field2:false,field3:3,field4:15}

const msgpack=require('@msgpack/msgpack')
const avscSchema=require('avsc').Type.forValue(data)
const packBytes=(...bitsVals)=>bitsVals.reduce((byte,val,i)=>i%2?byte+val:byte<<val,0)>>>0

const encodeFns={
   packBytes:data=>{
      const buf=Buffer.allocUnsafe(1)
      buf.writeUInt8(packBytes(1,data.field1,1,data.field2,2,data.field3,4,data.field4))
      return buf
   },
   avsc:data=>avscSchema.toBuffer(data),
   msgpack:data=>msgpack.encode(data),
   json:data=>Buffer.from(JSON.stringify(data)),
}

// node benchmark.js [encodings=all] [iterations=2_000_000]
const encodings=process.argv[2]&&process.argv[2]!='all'?[process.argv[2]]:Object.keys(encodeFns)
const iterations=process.argv[3]||2_000_000

encodings.forEach(type=>{
   let i=iterations,time=process.hrtime.bigint()
   while(i--)encodeFns[type](data)
   console.log(`${type} encode time(ns): ${Math.floor(Number(process.hrtime.bigint()-time)/iterations)} - bytes: ${encodeFns[type](data).length}`)
})
```

Encoding | time (ns) | bytes
--- | --- | ---
`packBytes` | 65 | 1
`avsc` | 235 | 4
`msgpack` | 1475 | 33
`json` | 765 | 53

**Reply by Wyzix33** [reply]
Hi, thank you for your answer, i see the benefits of using packBytes, i will use it.

**Reply by Wyzix33** [reply]
On loading it sends creature data, about 20000 creatures, 15000 spells and player info from a  MongoDB, it looks like this: 
```
[
 {
  "_id": "610266540a5ed77044b9d8b0",
  "creatureName": "Medusa",
  "hp": 2000000,
  "level": 60,
  "mainSpell": { "type": 14412, "damage": [10, 30] },
  "types": ["Boss", "Helper"],
  "desc": "Medusa, known for having hair made of snakes and for her ability to turn anyone she looked at to stone, literally to petrify.",
  "locations": [
   { "location": [32.53421523624556, 22.76575634523445], "spawnTime": 300 },
   { "location": [53.75734563434452, 66.53435341134533], "spawnTime": 300 }
  ]
 },
 {
  "_id": "610266540a5ed77044b9d8d1",
  "creatureName": "Grendel",
  "hp": 130000,
  "mainSpell": { "type": 66453, "damage": [4, 17] },
  "level": 20,
  "types": ["NPC"],
  "desc": "Grendel is the monster from the Old English epic poem Beowulf, and no one’s really sure of his true nature.",
  "locations": [
   { "location": [65.67111111111112, 95.17492654697409], "spawnTime": 300 },
   { "location": [15.53235523145212, 65.53422345326234], "spawnTime": 300 }
  ]
 }
]

```
// player info 
```
{
 "_id": "622341d2b3c2e8b593ba235f",
 "userid": "621b61aaee5803d360fa581c",
 "type": "Dwarf",
 "name": "Renown the Patient",
 "location": [65.67111111111112, 95.17492654697409],
 "level": 33,
 "hp": 92,
 "power": 10,
 "stamina": 30,
 "focus": 120,
 "agility": 11,
 "crit": 140,
 "haste": 22,
 "mastery": 102,
 "pets": [
  { "name": "Spirit Beast", "type": "wolf", "hp": 92, "power": 10, "stamina": 30 },
  { "name": "Doc", "type": "monkey", "hp": 122, "power": 30, "stamina": 60 }
 ]
}
```
// spells
```
[
 {
  "_id": "621b61aaee5803d360fa5830",
  "name": "Fireball",
  "spell_id": 2001,
  "cost": 14,
  "costType": "mana",
  "range": 31,
  "castTime": 1.5,
  "cooldown": 1,
  "effects": [44, 153],
  "duration": 4,
  "school": "fire"
 },
 {
  "_id": "621b61aaee5803d360fa5837",
  "name": "Firebolt",
  "spell_id": 2002,
  "cost": 10 ,
  "costType": "mana",
  "range": 40,
  "castTime": 3,
  "effects": [200],
  "school": "fire"
 }
]
```
Some spells don't have all the properties... but they can be 0 or null

After the browser receives all this the rest is really simple, the browser only sends to the server simple objects with player location update and spells / chat / actions like this:
```
{ "type": "move", "value": [65.67111111111112, 95.17492654697409] }
{ "type": "say", "value": "Hello world" }
{ "type": "cast", "value": { "spell_id": 2231, "selected": "creature", "_id": "610266540a5ed77044b9d8d1" } }
{ "type": "cast", "value": { "spell_id": 113, "selected": "player", "_id": "622341d2b3c2e8b593ba235f" } }
{ "type": "cast", "value": { "spell_id": 113, "selected": "self" } }
```
after calculating some values and checking if the action is possible the server broadcast to all other players the updates 

I understand that uWS has nothing to do with encoding, but any help is much appreciated :) and maybe will help others too
I agree with @e3dio that using vanilla is way better, I'll try to create a helper class or function that will take a object and return a buffer and the bit sizes that i need to send on the other side to unpack, and also need to find a solution for the array or object values {"a": [1,2]} or {"a":{"b":"c"}} and some date/time/timestamps fields.
Thanks

**Comment by ghost** [comment]
Just a side note - encoding/decoding your application's data has no relation to uWS. It is entirely disjoint, isolated from uWS.

**Reply by e3dio** [reply]
This discussion question belongs in the "avsc" repo on how to use that library. You don't need to use any external libraries though, better to use vanilla Node custom encode
